kvclient,sql: address a couple of TODOs for buffered writes

yuzefovich opened the pull request:
**kvclient: improve mockLockedSender**

Previously, `NumCalled` would only keep track of the number of times the mock function of mockLockedSender has been called. In particular, this would mean that all `SendLocked` calls before the first `MockSend` in a test would not count towards the `NumCalled` number. I find this behavior confusing, so we now will track all `SendLocked` calls.

**kvclient,sql: address a couple of TODOs for buffered writes**

This commit addresses a couple of TODOs:
- we now actually flush the buffer when write buffering goes from "enabled" to "disabled" (which is done when we see the first DDL on the txn). In order to simplify the logic, the flush occurs on the next BatchRequest.
- we now only enable write buffering on the txn under serializable isolation. Read committed has a known issue, but snapshot _should_ be ok, yet we choose to be safe and disable for both.

Epic: None
Release note: None

yuzefovich committed:
kvclient,sql: address a couple of TODOs for buffered writes

This commit addresses a couple of TODOs:
- we now actually flush the buffer when write buffering goes from
"enabled" to "disabled" (which is done when we see the first DDL on
the txn). In order to simplify the logic, the flush occurs on the next
BatchRequest.
- we now only enable write buffering on the txn under serializable
isolation. Read committed has a known issue, but snapshot _should_ be
ok, yet we choose to be safe and disable for both.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144171)
<!-- Reviewable:end -->

yuzefovich committed:
kvclient: improve mockLockedSender

Previously, `NumCalled` would only keep track of the number of times the
mock function of mockLockedSender has been called. In particular, this
would mean that all `SendLocked` calls before the first `MockSend` in
a test would not count towards the `NumCalled` number. I find this
behavior confusing, so we now will track all `SendLocked` calls.

Release note: None

yuzefovich reviewed (commented):
Added a test, RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144171#-:-ONRFa8z2YeDyWOPh6yk:b-cnpk1n)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @stevendanna)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144171#-:-ONRNYglD9-PhfIRHkeA:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @stevendanna)

___
*[`-- commits` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144171#-ONRMPZXCJNfllgXDxi8:-ONRMPZXCJNfllgXDxi9:bgxdl6h):*
nit: missing period.
___
*[`pkg/kv/kvclient/kvcoord/txn_coord_sender.go` line 1170 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144171#-ONRNEtP4pz_iQIakQ8h:-ONRNEtP4pz_iQIakQ8i:b74wen9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72970c422aabf7f968f88f663fd5da8814cd1a2/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L1170)):*
> ```Go
> 	defer tc.mu.Unlock()
> 
> 	if enabled == tc.interceptorAlloc.txnWriteBuffer.enabled {
> ```

Consider pulling this into a function on the `txnWriteBuffer`.  Then, you can just call it instead of trying to simulate it in the test below.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 1782 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144171#-ONRNRPB6Tuds5HE4QDw:-ONRNRPC8gkvBmRnBWo3:b-esg09e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72970c422aabf7f968f88f663fd5da8814cd1a2/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L1782)):*
> ```Go
> 	// Ensure the buffer is empty at this point.
> 	require.Equal(t, 0, len(twb.testingBufferedWritesAsSlice()))
> ```

For good measure, we could send another request to ensure the buffer remains empty.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144171#-:-ONRQkDg-_7AsKQaoe-S:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/txn_coord_sender.go` line 1170 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144171#-ONRNEtP4pz_iQIakQ8h:-ONRP6Xj1kSjGtVvaBF0:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72970c422aabf7f968f88f663fd5da8814cd1a2/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L1170)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Consider pulling this into a function on the `txnWriteBuffer`.  Then, you can just call it instead of trying to simulate it in the test below.
</blockquote></details>

Good point, done.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 1782 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144171#-ONRNRPB6Tuds5HE4QDw:-ONRQT4Y9u9i85eMVXwB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72970c422aabf7f968f88f663fd5da8814cd1a2/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L1782)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

For good measure, we could send another request to ensure the buffer remains empty.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r+

yuzefovich commented:
bors r+
